README.md
# maxwell-protocol-dart

CHANGELOG.md
# Changelog

This project follows [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html).

## [0.7.1] - 2023-11-16

### Changed

- Bumped dependencies


test/maxwell_protocol_test.dart
import 'package:test/test.dart';
import 'package:maxwell_protocol/maxwell_protocol.dart';

void main() {
  test("req_req_t", () {
    var header = header_t();
    header.agent = "chrome";
    header.endpoint = "192.168.1.1:8080";
    var req = req_req_t();
    req.path = "get_candles";
    req.payload = "{}";
    req.header = header;
    req.ref = 1;
    expect(req.header.agent, equals("chrome"));
    var encodedReq = encode_msg(req);
    req_req_t req2 = decode_msg(encodedReq) as req_req_t;
    expect(req == req2, true);
    expect(req2.ref, 99);
    expect(req.get_ref(), 1);
  });
}


lib/src/maxwell_protocol.pbenum.dart
//
//  Generated code. Do not modify.
//  source: maxwell_protocol.proto
//
// @dart = 2.12

// ignore_for_file: annotate_overrides, camel_case_types
// ignore_for_file: constant_identifier_names, library_prefixes
// ignore_for_file: non_constant_identifier_names, prefer_final_fields
// ignore_for_file: unnecessary_import, unnecessary_this, unused_import

import 'dart:core' as $core;

import 'package:protobuf/protobuf.dart' as $pb;

class msg_type_t extends $pb.ProtobufEnum {
  static const msg_type_t UNKNOWN = msg_type_t._(0, _omitEnumNames ? '' : 'UNKNOWN');
  static const msg_type_t PING_REQ = msg_type_t._(1, _omitEnumNames ? '' : 'PING_REQ');
  static const msg_type_t PING_REP = msg_type_t._(2, _omitEnumNames ? '' : 'PING_REP');
  static const msg_type_t OK_REP = msg_type_t._(29, _omitEnumNames ? '' : 'OK_REP');
  static const msg_type_t ERROR_REP = msg_type_t._(30, _omitEnumNames ? '' : 'ERROR_REP');
  static const msg_type_t OK2_REP = msg_type_t._(31, _omitEnumNames ? '' : 'OK2_REP');
  static const msg_type_t ERROR2_REP = msg_type_t._(32, _omitEnumNames ? '' : 'ERROR2_REP');
  static const msg_type_t PUSH_REQ = msg_type_t._(33, _omitEnumNames ? '' : 'PUSH_REQ');
  static const msg_type_t PUSH_REP = msg_type_t._(34, _omitEnumNames ? '' : 'PUSH_REP');
  static const msg_type_t PULL_REQ = msg_type_t._(35, _omitEnumNames ? '' : 'PULL_REQ');
  static const msg_type_t PULL_REP = msg_type_t._(36, _omitEnumNames ? '' : 'PULL_REP');
  static const msg_type_t REQ_REQ = msg_type_t._(39, _omitEnumNames ? '' : 'REQ_REQ');
  static const msg_type_t REQ_REP = msg_type_t._(40, _omitEnumNames ? '' : 'REQ_REP');
  static const msg_type_t AUTH_REQ = msg_type_t._(41, _omitEnumNames ? '' : 'AUTH_REQ');
  static const msg_type_t AUTH_REP = msg_type_t._(42, _omitEnumNames ? '' : 'AUTH_REP');
  static const msg_type_t REGISTER_FRONTEND_REQ = msg_type_t._(65, _omitEnumNames ? '' : 'REGISTER_FRONTEND_REQ');
  static const msg_type_t REGISTER_FRONTEND_REP = msg_type_t._(66, _omitEnumNames ? '' : 'REGISTER_FRONTEND_REP');
  static const msg_type_t REGISTER_BACKEND_REQ = msg_type_t._(67, _omitEnumNames ? '' : 'REGISTER_BACKEND_REQ');
  static const msg_type_t REGISTER_BACKEND_REP = msg_type_t._(68, _omitEnumNames ? '' : 'REGISTER_BACKEND_REP');
  static const msg_type_t REGISTER_SERVICE_REQ = msg_type_t._(69, _omitEnumNames ? '' : 'REGISTER_SERVICE_REQ');
  static const msg_type_t REGISTER_SERVICE_REP = msg_type_t._(70, _omitEnumNames ? '' : 'REGISTER_SERVICE_REP');
  static const msg_type_t SET_ROUTES_REQ = msg_type_t._(71, _omitEnumNames ? '' : 'SET_ROUTES_REQ');
  static const msg_type_t SET_ROUTES_REP = msg_type_t._(72, _omitEnumNames ? '' : 'SET_ROUTES_REP');
  static const msg_type_t GET_ROUTES_REQ = msg_type_t._(75, _omitEnumNames ? '' : 'GET_ROUTES_REQ');
  static const msg_type_t GET_ROUTES_REP = msg_type_t._(76, _omitEnumNames ? '' : 'GET_ROUTES_REP');
  static const msg_type_t GET_TOPIC_DIST_CHECKSUM_REQ = msg_type_t._(77, _omitEnumNames ? '' : 'GET_TOPIC_DIST_CHECKSUM_REQ');
  static const msg_type_t GET_TOPIC_DIST_CHECKSUM_REP = msg_type_t._(78, _omitEnumNames ? '' : 'GET_TOPIC_DIST_CHECKSUM_REP');
  static const msg_type_t GET_ROUTE_DIST_CHECKSUM_REQ = msg_type_t._(79, _omitEnumNames ? '' : 'GET_ROUTE_DIST_CHECKSUM_REQ');
  static const msg_type_t GET_ROUTE_DIST_CHECKSUM_REP = msg_type_t._(80, _omitEnumNames ? '' : 'GET_ROUTE_DIST_CHECKSUM_REP');
  static const msg_type_t PICK_FRONTEND_REQ = msg_type_t._(81, _omitEnumNames ? '' : 'PICK_FRONTEND_REQ');
  static const msg_type_t PICK_FRONTEND_REP = msg_type_t._(82, _omitEnumNames ? '' : 'PICK_FRONTEND_REP');
  static const msg_type_t PICK_FRONTENDS_REQ = msg_type_t._(83, _omitEnumNames ? '' : 'PICK_FRONTENDS_REQ');
  static const msg_type_t PICK_FRONTENDS_REP = msg_type_t._(84, _omitEnumNames ? '' : 'PICK_FRONTENDS_REP');
  static const msg_type_t LOCATE_TOPIC_REQ = msg_type_t._(85, _omitEnumNames ? '' : 'LOCATE_TOPIC_REQ');
  static const msg_type_t LOCATE_TOPIC_REP = msg_type_t._(86, _omitEnumNames ? '' : 'LOCATE_TOPIC_REP');
  static const msg_type_t RESOLVE_IP_REQ = msg_type_t._(121, _omitEnumNames ? '' : 'RESOLVE_IP_REQ');
  static const msg_type_t RESOLVE_IP_REP = msg_type_t._(122, _omitEnumNames ? '' : 'RESOLVE_IP_REP');

  static const $core.List<msg_type_t> values = <msg_type_t> [
    UNKNOWN,
    PING_REQ,
    PING_REP,
    OK_REP,
    ERROR_REP,
    OK2_REP,
    ERROR2_REP,
    PUSH_REQ,
    PUSH_REP,
    PULL_REQ,
    PULL_REP,
    REQ_REQ,
    REQ_REP,
    AUTH_REQ,
    AUTH_REP,
    REGISTER_FRONTEND_REQ,
    REGISTER_FRONTEND_REP,
    REGISTER_BACKEND_REQ,
    REGISTER_BACKEND_REP,
    REGISTER_SERVICE_REQ,
    REGISTER_SERVICE_REP,
    SET_ROUTES_REQ,
    SET_ROUTES_REP,
    GET_ROUTES_REQ,
    GET_ROUTES_REP,
    GET_TOPIC_DIST_CHECKSUM_REQ,
    GET_TOPIC_DIST_CHECKSUM_REP,
    GET_ROUTE_DIST_CHECKSUM_REQ,
    GET_ROUTE_DIST_CHECKSUM_REP,
    PICK_FRONTEND_REQ,
    PICK_FRONTEND_REP,
    PICK_FRONTENDS_REQ,
    PICK_FRONTENDS_REP,
    LOCATE_TOPIC_REQ,
    LOCATE_TOPIC_REP,
    RESOLVE_IP_REQ,
    RESOLVE_IP_REP,
  ];

  static final $core.Map<$core.int, msg_type_t> _byValue = $pb.ProtobufEnum.initByValue(values);
  static msg_type_t? valueOf($core.int value) => _byValue[value];

  const msg_type_t._($core.int v, $core.String n) : super(v, n);
}

class error_code_t extends $pb.ProtobufEnum {
  static const error_code_t OK = error_code_t._(0, _omitEnumNames ? '' : 'OK');
  static const error_code_t UNKNOWN_MSG = error_code_t._(1, _omitEnumNames ? '' : 'UNKNOWN_MSG');
  static const error_code_t NOT_ALLOWED_TO_REGISTER_FRONTEND = error_code_t._(100, _omitEnumNames ? '' : 'NOT_ALLOWED_TO_REGISTER_FRONTEND');
  static const error_code_t NOT_ALLOWED_TO_REGISTER_BACKEND = error_code_t._(101, _omitEnumNames ? '' : 'NOT_ALLOWED_TO_REGISTER_BACKEND');
  static const error_code_t NOT_ALLOWED_TO_REGISTER_SERVICE = error_code_t._(102, _omitEnumNames ? '' : 'NOT_ALLOWED_TO_REGISTER_SERVICE');
  static const error_code_t FAILED_TO_SET_ROUTES = error_code_t._(103, _omitEnumNames ? '' : 'FAILED_TO_SET_ROUTES');
  static const error_code_t FAILED_TO_PICK_FRONTEND = error_code_t._(104, _omitEnumNames ? '' : 'FAILED_TO_PICK_FRONTEND');
  static const error_code_t FAILED_TO_LOCATE_TOPIC = error_code_t._(105, _omitEnumNames ? '' : 'FAILED_TO_LOCATE_TOPIC');
  static const error_code_t MASTER_ERROR = error_code_t._(199, _omitEnumNames ? '' : 'MASTER_ERROR');
  static const error_code_t FAILED_TO_REQUEST_SERVICE = error_code_t._(200, _omitEnumNames ? '' : 'FAILED_TO_REQUEST_SERVICE');
  static const error_code_t FAILED_TO_REQUEST_BACKEND = error_code_t._(201, _omitEnumNames ? '' : 'FAILED_TO_REQUEST_BACKEND');
  static const error_code_t FRONTEND_ERROR = error_code_t._(299, _omitEnumNames ? '' : 'FRONTEND_ERROR');
  static const error_code_t FAILED_TO_PUSH = error_code_t._(300, _omitEnumNames ? '' : 'FAILED_TO_PUSH');
  static const error_code_t FAILED_TO_PULL = error_code_t._(301, _omitEnumNames ? '' : 'FAILED_TO_PULL');
  static const error_code_t UNKNOWN_TOPIC = error_code_t._(302, _omitEnumNames ? '' : 'UNKNOWN_TOPIC');
  static const error_code_t BACKEND_ERROR = error_code_t._(399, _omitEnumNames ? '' : 'BACKEND_ERROR');
  static const error_code_t UNKNOWN_PATH = error_code_t._(400, _omitEnumNames ? '' : 'UNKNOWN_PATH');
  static const error_code_t SERVICE_ERROR = error_code_t._(499, _omitEnumNames ? '' : 'SERVICE_ERROR');
  static const error_code_t CLIENT_ERROR = error_code_t._(599, _omitEnumNames ? '' : 'CLIENT_ERROR');

  static const $core.List<error_code_t> values = <error_code_t> [
    OK,
    UNKNOWN_MSG,
    NOT_ALLOWED_TO_REGISTER_FRONTEND,
    NOT_ALLOWED_TO_REGISTER_BACKEND,
    NOT_ALLOWED_TO_REGISTER_SERVICE,
    FAILED_TO_SET_ROUTES,
    FAILED_TO_PICK_FRONTEND,
    FAILED_TO_LOCATE_TOPIC,
    MASTER_ERROR,
    FAILED_TO_REQUEST_SERVICE,
    FAILED_TO_REQUEST_BACKEND,
    FRONTEND_ERROR,
    FAILED_TO_PUSH,
    FAILED_TO_PULL,
    UNKNOWN_TOPIC,
    BACKEND_ERROR,
    UNKNOWN_PATH,
    SERVICE_ERROR,
    CLIENT_ERROR,
  ];

  static final $core.Map<$core.int, error_code_t> _byValue = $pb.ProtobufEnum.initByValue(values);
  static error_code_t? valueOf($core.int value) => _byValue[value];

  const error_code_t._($core.int v, $core.String n) : super(v, n);
}


const _omitEnumNames = $core.bool.fromEnvironment('protobuf.omit_enum_names');


lib/maxwell_protocol.dart
import 'dart:typed_data';
import 'package:protobuf/protobuf.dart';
import 'package:maxwell_protocol/src/maxwell_protocol.pb.dart';

export 'package:protobuf/protobuf.dart';
export 'package:maxwell_protocol/src/maxwell_protocol.pb.dart';

Uint8List encode_msg(GeneratedMessage msg) {
  var encodedMsgType = Uint8List(1);
  var encodedMsgBody = msg.writeToBuffer();
  switch(msg.runtimeType) {
  case auth_rep_t: {
    encodedMsgType[0] = 42;
    break;
  }
  case auth_req_t: {
    encodedMsgType[0] = 41;
    break;
  }
  case error2_rep_t: {
    encodedMsgType[0] = 32;
    break;
  }
  case error_rep_t: {
    encodedMsgType[0] = 30;
    break;
  }
  case get_routes_rep_t: {
    encodedMsgType[0] = 76;
    break;
  }
  case get_routes_req_t: {
    encodedMsgType[0] = 75;
    break;
  }
  case get_route_dist_checksum_rep_t: {
    encodedMsgType[0] = 80;
    break;
  }
  case get_route_dist_checksum_req_t: {
    encodedMsgType[0] = 79;
    break;
  }
  case get_topic_dist_checksum_rep_t: {
    encodedMsgType[0] = 78;
    break;
  }
  case get_topic_dist_checksum_req_t: {
    encodedMsgType[0] = 77;
    break;
  }
  case locate_topic_rep_t: {
    encodedMsgType[0] = 86;
    break;
  }
  case locate_topic_req_t: {
    encodedMsgType[0] = 85;
    break;
  }
  case ok2_rep_t: {
    encodedMsgType[0] = 31;
    break;
  }
  case ok_rep_t: {
    encodedMsgType[0] = 29;
    break;
  }
  case pick_frontends_rep_t: {
    encodedMsgType[0] = 84;
    break;
  }
  case pick_frontends_req_t: {
    encodedMsgType[0] = 83;
    break;
  }
  case pick_frontend_rep_t: {
    encodedMsgType[0] = 82;
    break;
  }
  case pick_frontend_req_t: {
    encodedMsgType[0] = 81;
    break;
  }
  case ping_rep_t: {
    encodedMsgType[0] = 2;
    break;
  }
  case ping_req_t: {
    encodedMsgType[0] = 1;
    break;
  }
  case pull_rep_t: {
    encodedMsgType[0] = 36;
    break;
  }
  case pull_req_t: {
    encodedMsgType[0] = 35;
    break;
  }
  case push_rep_t: {
    encodedMsgType[0] = 34;
    break;
  }
  case push_req_t: {
    encodedMsgType[0] = 33;
    break;
  }
  case register_backend_rep_t: {
    encodedMsgType[0] = 68;
    break;
  }
  case register_backend_req_t: {
    encodedMsgType[0] = 67;
    break;
  }
  case register_frontend_rep_t: {
    encodedMsgType[0] = 66;
    break;
  }
  case register_frontend_req_t: {
    encodedMsgType[0] = 65;
    break;
  }
  case register_service_rep_t: {
    encodedMsgType[0] = 70;
    break;
  }
  case register_service_req_t: {
    encodedMsgType[0] = 69;
    break;
  }
  case req_rep_t: {
    encodedMsgType[0] = 40;
    break;
  }
  case req_req_t: {
    encodedMsgType[0] = 39;
    break;
  }
  case resolve_ip_rep_t: {
    encodedMsgType[0] = 122;
    break;
  }
  case resolve_ip_req_t: {
    encodedMsgType[0] = 121;
    break;
  }
  case set_routes_rep_t: {
    encodedMsgType[0] = 72;
    break;
  }
  case set_routes_req_t: {
    encodedMsgType[0] = 71;
    break;
  }
  default: {
    throw 'Unknown msg type: ${msg.runtimeType}';
  }}
  return Uint8List.fromList(encodedMsgType + encodedMsgBody);
}

GeneratedMessage decode_msg(Uint8List encodedMsg) {
  var msgType = encodedMsg[0];
  var encodedMsgBody = encodedMsg.sublist(1);
  switch(msgType) {
  case 42: {
    return auth_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 41: {
    return auth_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 32: {
    return error2_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 30: {
    return error_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 76: {
    return get_routes_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 75: {
    return get_routes_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 80: {
    return get_route_dist_checksum_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 79: {
    return get_route_dist_checksum_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 78: {
    return get_topic_dist_checksum_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 77: {
    return get_topic_dist_checksum_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 86: {
    return locate_topic_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 85: {
    return locate_topic_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 31: {
    return ok2_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 29: {
    return ok_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 84: {
    return pick_frontends_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 83: {
    return pick_frontends_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 82: {
    return pick_frontend_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 81: {
    return pick_frontend_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 2: {
    return ping_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 1: {
    return ping_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 36: {
    return pull_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 35: {
    return pull_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 34: {
    return push_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 33: {
    return push_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 68: {
    return register_backend_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 67: {
    return register_backend_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 66: {
    return register_frontend_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 65: {
    return register_frontend_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 70: {
    return register_service_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 69: {
    return register_service_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 40: {
    return req_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 39: {
    return req_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 122: {
    return resolve_ip_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 121: {
    return resolve_ip_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 72: {
    return set_routes_rep_t()..mergeFromBuffer(encodedMsgBody);
  }
  case 71: {
    return set_routes_req_t()..mergeFromBuffer(encodedMsgBody);
  }
  default: {
    throw 'Unknown msg type: ${msgType}';
  }}
}

extension ref_aware on GeneratedMessage {
  int get_ref() {
    switch(this.runtimeType) {
    case auth_rep_t: {
      return (this as auth_rep_t).ref;
    }
    case auth_req_t: {
      return (this as auth_req_t).ref;
    }
    case error2_rep_t: {
      return (this as error2_rep_t).ref;
    }
    case error_rep_t: {
      return (this as error_rep_t).ref;
    }
    case get_routes_rep_t: {
      return (this as get_routes_rep_t).ref;
    }
    case get_routes_req_t: {
      return (this as get_routes_req_t).ref;
    }
    case get_route_dist_checksum_rep_t: {
      return (this as get_route_dist_checksum_rep_t).ref;
    }
    case get_route_dist_checksum_req_t: {
      return (this as get_route_dist_checksum_req_t).ref;
    }
    case get_topic_dist_checksum_rep_t: {
      return (this as get_topic_dist_checksum_rep_t).ref;
    }
    case get_topic_dist_checksum_req_t: {
      return (this as get_topic_dist_checksum_req_t).ref;
    }
    case locate_topic_rep_t: {
      return (this as locate_topic_rep_t).ref;
    }
    case locate_topic_req_t: {
      return (this as locate_topic_req_t).ref;
    }
    case ok2_rep_t: {
      return (this as ok2_rep_t).ref;
    }
    case ok_rep_t: {
      return (this as ok_rep_t).ref;
    }
    case pick_frontends_rep_t: {
      return (this as pick_frontends_rep_t).ref;
    }
    case pick_frontends_req_t: {
      return (this as pick_frontends_req_t).ref;
    }
    case pick_frontend_rep_t: {
      return (this as pick_frontend_rep_t).ref;
    }
    case pick_frontend_req_t: {
      return (this as pick_frontend_req_t).ref;
    }
    case ping_rep_t: {
      return (this as ping_rep_t).ref;
    }
    case ping_req_t: {
      return (this as ping_req_t).ref;
    }
    case pull_rep_t: {
      return (this as pull_rep_t).ref;
    }
    case pull_req_t: {
      return (this as pull_req_t).ref;
    }
    case push_rep_t: {
      return (this as push_rep_t).ref;
    }
    case push_req_t: {
      return (this as push_req_t).ref;
    }
    case register_backend_rep_t: {
      return (this as register_backend_rep_t).ref;
    }
    case register_backend_req_t: {
      return (this as register_backend_req_t).ref;
    }
    case register_frontend_rep_t: {
      return (this as register_frontend_rep_t).ref;
    }
    case register_frontend_req_t: {
      return (this as register_frontend_req_t).ref;
    }
    case register_service_rep_t: {
      return (this as register_service_rep_t).ref;
    }
    case register_service_req_t: {
      return (this as register_service_req_t).ref;
    }
    case req_rep_t: {
      return (this as req_rep_t).ref;
    }
    case req_req_t: {
      return (this as req_req_t).ref;
    }
    case resolve_ip_rep_t: {
      return (this as resolve_ip_rep_t).ref;
    }
    case resolve_ip_req_t: {
      return (this as resolve_ip_req_t).ref;
    }
    case set_routes_rep_t: {
      return (this as set_routes_rep_t).ref;
    }
    case set_routes_req_t: {
      return (this as set_routes_req_t).ref;
    }
    default: {
      throw 'Unknown msg type: ${this.runtimeType}';
    }}
  }

  void set_ref(int ref) {
    switch(this.runtimeType) {
      case auth_rep_t: {
      (this as auth_rep_t).ref = ref;
      break;
    }
    case auth_req_t: {
      (this as auth_req_t).ref = ref;
      break;
    }
    case error2_rep_t: {
      (this as error2_rep_t).ref = ref;
      break;
    }
    case error_rep_t: {
      (this as error_rep_t).ref = ref;
      break;
    }
    case get_routes_rep_t: {
      (this as get_routes_rep_t).ref = ref;
      break;
    }
    case get_routes_req_t: {
      (this as get_routes_req_t).ref = ref;
      break;
    }
    case get_route_dist_checksum_rep_t: {
      (this as get_route_dist_checksum_rep_t).ref = ref;
      break;
    }
    case get_route_dist_checksum_req_t: {
      (this as get_route_dist_checksum_req_t).ref = ref;
      break;
    }
    case get_topic_dist_checksum_rep_t: {
      (this as get_topic_dist_checksum_rep_t).ref = ref;
      break;
    }
    case get_topic_dist_checksum_req_t: {
      (this as get_topic_dist_checksum_req_t).ref = ref;
      break;
    }
    case locate_topic_rep_t: {
      (this as locate_topic_rep_t).ref = ref;
      break;
    }
    case locate_topic_req_t: {
      (this as locate_topic_req_t).ref = ref;
      break;
    }
    case ok2_rep_t: {
      (this as ok2_rep_t).ref = ref;
      break;
    }
    case ok_rep_t: {
      (this as ok_rep_t).ref = ref;
      break;
    }
    case pick_frontends_rep_t: {
      (this as pick_frontends_rep_t).ref = ref;
      break;
    }
    case pick_frontends_req_t: {
      (this as pick_frontends_req_t).ref = ref;
      break;
    }
    case pick_frontend_rep_t: {
      (this as pick_frontend_rep_t).ref = ref;
      break;
    }
    case pick_frontend_req_t: {
      (this as pick_frontend_req_t).ref = ref;
      break;
    }
    case ping_rep_t: {
      (this as ping_rep_t).ref = ref;
      break;
    }
    case ping_req_t: {
      (this as ping_req_t).ref = ref;
      break;
    }
    case pull_rep_t: {
      (this as pull_rep_t).ref = ref;
      break;
    }
    case pull_req_t: {
      (this as pull_req_t).ref = ref;
      break;
    }
    case push_rep_t: {
      (this as push_rep_t).ref = ref;
      break;
    }
    case push_req_t: {
      (this as push_req_t).ref = ref;
      break;
    }
    case register_backend_rep_t: {
      (this as register_backend_rep_t).ref = ref;
      break;
    }
    case register_backend_req_t: {
      (this as register_backend_req_t).ref = ref;
      break;
    }
    case register_frontend_rep_t: {
      (this as register_frontend_rep_t).ref = ref;
      break;
    }
    case register_frontend_req_t: {
      (this as register_frontend_req_t).ref = ref;
      break;
    }
    case register_service_rep_t: {
      (this as register_service_rep_t).ref = ref;
      break;
    }
    case register_service_req_t: {
      (this as register_service_req_t).ref = ref;
      break;
    }
    case req_rep_t: {
      (this as req_rep_t).ref = ref;
      break;
    }
    case req_req_t: {
      (this as req_req_t).ref = ref;
      break;
    }
    case resolve_ip_rep_t: {
      (this as resolve_ip_rep_t).ref = ref;
      break;
    }
    case resolve_ip_req_t: {
      (this as resolve_ip_req_t).ref = ref;
      break;
    }
    case set_routes_rep_t: {
      (this as set_routes_rep_t).ref = ref;
      break;
    }
    case set_routes_req_t: {
      (this as set_routes_req_t).ref = ref;
      break;
    }
    default: {
      throw 'Unknown msg type: ${this.runtimeType}';
    }}
  }
}